storage/redis/peer_store.go

```diff
@@ -0,0 +1,826 @@
+				"key":   ps.leecherCountKey(group),
+				"error": err,
+			})
+		} else {
+			numLeechers += n.(int64)
+		}
+	}
+
+	storage.PromInfohashesCount.Set(float64(numInfohashes))
+	storage.PromSeedersCount.Set(float64(numSeeders))
+	storage.PromLeechersCount.Set(float64(numLeechers))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return timecache.NowUnixNano()
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: PutSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p,
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	encodedSeederInfoHash := ps.seederInfohashKey(addressFamily, ih.String())
+	ct := ps.getClock()
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	conn.Send("MULTI")
+	conn.Send("HSET", encodedSeederInfoHash, pk, ct)
+	conn.Send("HSET", addressFamily, encodedSeederInfoHash, ct)
+	reply, err := redis.Values(conn.Do("EXEC"))
+	if err != nil {
+		return err
+	}
+
+	// pk is a new field.
+	if reply[0].(int64) == 1 {
+		_, err = conn.Do("INCR", ps.seederCountKey(addressFamily))
+		if err != nil {
+			return err
+		}
+	}
+	// encodedSeederInfoHash is a new field.
+	if reply[1].(int64) == 1 {
+		_, err = conn.Do("INCR", ps.infohashCountKey(addressFamily))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) DeleteSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: DeleteSeeder", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p,
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encodedSeederInfoHash := ps.seederInfohashKey(addressFamily, ih.String())
+
+	delNum, err := conn.Do("HDEL", encodedSeederInfoHash, pk)
+	if err != nil {
+		return err
+	}
+	if delNum.(int64) == 0 {
+		return storage.ErrResourceDoesNotExist
+	}
+	if _, err := conn.Do("DECR", ps.seederCountKey(addressFamily)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) PutLeecher(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	addressFamily := p.IP.AddressFamily.String()
+	log.Debug("storage: PutLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     p,
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	// Update the peer in the swarm.
+	encodedLeecherInfoHash := ps.leecherInfohashKey(addressFamily, ih.String())
+	pk := newPeerKey(p)
+	ct := ps.getClock()
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	conn.Send("MULTI")
+	conn.Send("HSET", encodedLeecherInfoHash, pk, ct)
+	conn.Send("HSET", addressFamily, encodedLeecherInfoHash, ct)
+	reply, err := redis.Values(conn.Do("EXEC"))
```

Comment:
This fails sometimes. I got it to fail by force re-checking the torrent in Deluge. What this does is it sends a `stopped` event to the tracker, then re-checks, then sends a `started` event to the tracker. This is the log:
```
DEBU[0318] sanitized announce                            1.defaultNumWant=50 1.maxNumWant=100 compact=true downloaded=0 event=stopped eventProvided=true infoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c ipProvided=false left=2799 numWant=0 numWantProvided=true params="&{/announce info_hash=%c8%81%fd%a6%e9ZGX%1dy%c0e%b7%0f%aa%1b%5d%e7%e9%9c&peer_id=-DE13D0-Z*YkFvQ~fnkP&port=60695&uploaded=0&downloaded=0&left=2799&corrupt=0&key=53A818CA&event=stopped&numwant=0&compact=1&no_peer_id=1&supportcrypto=1&redundant=0 map[downloaded:0 left:2799 event:stopped numwant:0 compact:1 supportcrypto:1 redundant:0 peer_id:-DE13D0-Z*YkFvQ~fnkP port:60695 uploaded:0 corrupt:0 key:53A818CA no_peer_id:1] [[200 129 253 166 233 90 71 88 29 121 192 101 183 15 170 27 93 231 233 156]]}" peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695" uploaded=0
2019/01/20 15:02:29 redis.Do(PING) -> ("PONG", <nil>)
2019/01/20 15:02:29 redis.Do(HLEN, "IPv4_L_c881fda6e95a47581d79c065b"...) -> (1, <nil>)
2019/01/20 15:02:29 redis.Do(HLEN, "IPv4_S_c881fda6e95a47581d79c065b"...) -> (0, <nil>)
2019/01/20 15:02:29 redis.Do() -> (<nil>, <nil>)
DEBU[0318] storage: AnnouncePeers                        InfoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c Peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695" numWant=0 seeder=false
2019/01/20 15:02:29 redis.Do(HKEYS, "IPv4_L_c881fda6e95a47581d79c065b"...) -> (["-DE13D0-Z*YkFvQ~fnkP\xed\x17\u007f\x00\x00\x01"], <nil>)
2019/01/20 15:02:29 redis.Do(HKEYS, "IPv4_S_c881fda6e95a47581d79c065b"...) -> ([], <nil>)
2019/01/20 15:02:29 redis.Do() -> (<nil>, <nil>)
DEBU[0318] generated announce response                   compact=true complete=0 interval=30m0s ipv4Peers="[2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695]" ipv6Peers="[]" minInterval=15m0s
DEBU[0318] storage: DeleteSeeder                         InfoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c Peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695"
2019/01/20 15:02:29 redis.Do(HDEL, "IPv4_S_c881fda6e95a47581d79c065b"..., -DE13D0-Z*YkFvQ~fnkP�) -> (0, <nil>)
2019/01/20 15:02:29 redis.Do() -> (<nil>, <nil>)
DEBU[0318] storage: DeleteLeecher                        InfoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c Peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695"
2019/01/20 15:02:29 redis.Do(HDEL, "IPv4_L_c881fda6e95a47581d79c065b"..., -DE13D0-Z*YkFvQ~fnkP�) -> (1, <nil>)
2019/01/20 15:02:29 redis.Do(DECR, "IPv4_L_count") -> (0, <nil>)
2019/01/20 15:02:29 redis.Do() -> (<nil>, <nil>)
DEBU[0318] sanitized announce                            1.defaultNumWant=50 1.maxNumWant=100 compact=true downloaded=0 event=started eventProvided=true infoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c ipProvided=false left=2799 numWant=100 numWantProvided=true params="&{/announce info_hash=%c8%81%fd%a6%e9ZGX%1dy%c0e%b7%0f%aa%1b%5d%e7%e9%9c&peer_id=-DE13D0-Z*YkFvQ~fnkP&port=60695&uploaded=0&downloaded=0&left=2799&corrupt=0&key=53A818CA&event=started&numwant=200&compact=1&no_peer_id=1&supportcrypto=1&redundant=0 map[downloaded:0 left:2799 key:53A818CA numwant:200 supportcrypto:1 port:60695 uploaded:0 event:started compact:1 no_peer_id:1 redundant:0 peer_id:-DE13D0-Z*YkFvQ~fnkP corrupt:0] [[200 129 253 166 233 90 71 88 29 121 192 101 183 15 170 27 93 231 233 156]]}" peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695" uploaded=0
2019/01/20 15:02:30 redis.Do(HLEN, "IPv4_L_c881fda6e95a47581d79c065b"...) -> (0, <nil>)
2019/01/20 15:02:30 redis.Do(HLEN, "IPv4_S_c881fda6e95a47581d79c065b"...) -> (0, <nil>)
2019/01/20 15:02:30 redis.Do() -> (<nil>, <nil>)
DEBU[0318] storage: AnnouncePeers                        InfoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c Peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695" numWant=100 seeder=false
2019/01/20 15:02:30 redis.Do(HKEYS, "IPv4_L_c881fda6e95a47581d79c065b"...) -> ([], <nil>)
2019/01/20 15:02:30 redis.Do(HKEYS, "IPv4_S_c881fda6e95a47581d79c065b"...) -> ([], <nil>)
2019/01/20 15:02:30 redis.Do() -> (<nil>, <nil>)
DEBU[0318] generated announce response                   compact=true complete=0 interval=30m0s ipv4Peers="[2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695]" ipv6Peers="[]" minInterval=15m0s
DEBU[0318] storage: PutLeecher                           InfoHash=c881fda6e95a47581d79c065b70faa1b5de7e99c Peer="2d4445313344302d5a2a596b4676517e666e6b50@[127.0.0.1]:60695"
2019/01/20 15:02:30 redis.Send(MULTI) -> (<nil>)
2019/01/20 15:02:30 redis.Send(HSET, "IPv4_L_c881fda6e95a47581d79c065b"..., -DE13D0-Z*YkFvQ~fnkP�, 1547985749810944415) -> (<nil>)
2019/01/20 15:02:30 redis.Send(HSET, "IPv4", "IPv4_L_c881fda6e95a47581d79c065b"..., 1547985749810944415) -> (<nil>)
2019/01/20 15:02:30 redis.Do(EXEC) -> (<nil>, <nil>)
2019/01/20 15:02:30 redis.Do() -> (<nil>, <nil>)
ERRO[0318] post-announce hooks failed                    error="redigo: nil returned"
```
I don't know why it fails, but it does.

It does not fail all the time.

EDIT: I think I figured this one out. It's because of the UNWATCH failing (because it has the wrong syntax). The key is still WATCHed and the EXEC here fails. That's actually kind of scary, because we WATCHed it in a different connection! I assumed connections were isolated with regard to WATCH MULTI EXEC transactions.

Reply:
This actually means that we have to wrap every `MULTI` ... `EXEC` block in a loop, because it could be failing from an unrelated 'WATCH` done in another connection.

Reply:
@mrd0ll4r Transaction is limited inside connection, WATCH/UNWATCH works in one connection, it doesn't have effects on other connections, I have experimented this by opening two `redis-cli`, the number represents the executing sequence

client-A
```
1. set name onestraw
2. watch name
6. multi
7. get name
8. set name larry.he
9. exec
(nil)
```
client-B
```
3. unwatch
4. set name one-starw
5. get name
```

Reply:
I agree, that was a mistake on my end. I think I saw the behavior because my requests were so slow and far apart that we simply recycled the same connection, and that still had the WATCH going. So yes, you're right, transactions are limited inside connections. :)